File: frontend/src/services/imageSearch.ts
import { api } from './api';
import type { Product } from './product';

export interface ImageSearchResponse {
  success: boolean;
  results: Product[];
}

export interface SimilarImageResponse {
  success: boolean;
  top_k: number;
  products: Product[];
}

export interface ReverseSearchResponse {
  success: boolean;
  matches: Array<{ product: Product; confidence: number }>;
}

export interface ImageRecommendResponse {
  success: boolean;
  recommendations: Product[];
}

export const imageSearchService = {
  async searchByImage(imageFile: File): Promise<ImageSearchResponse> {
    const formData = new FormData();
    formData.append('image', imageFile);
    const res = await api.post<ImageSearchResponse>('/api/image-search/search', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  },

  async findSimilarProducts(imageFile: File, topK = 5): Promise<SimilarImageResponse> {
    const formData = new FormData();
    formData.append('image', imageFile);
    const res = await api.post<SimilarImageResponse>('/api/image-search/similar', formData, {
      params: { top_k: topK },
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  },

  async reverseImageSearch(imageFile: File): Promise<ReverseSearchResponse> {
    const formData = new FormData();
    formData.append('image', imageFile);
    const res = await api.post<ReverseSearchResponse>('/api/image-search/reverse', formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  },

  async recommendByImage(imageFile: File, topK = 10): Promise<ImageRecommendResponse> {
    const formData = new FormData();
    formData.append('image', imageFile);
    const res = await api.post<ImageRecommendResponse>('/api/image-search/recommend', formData, {
      params: { top_k: topK },
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return res.data;
  }
};

File: backend/app/api/image_search.py
from fastapi import APIRouter, UploadFile, File, HTTPException
from pydantic import BaseModel
from typing import List

from app.services.image_search_service import ImageSearchService

router = APIRouter()

image_service = ImageSearchService()


# ---------------- Image Upload Search ----------------
@router.post("/search")
async def image_search(
        image: UploadFile = File(...)
):
    """
    Search products using uploaded image.
    """

    try:
        result = await image_service.search_by_image(image)

        return {
            "success": True,
            "results": result
        }

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=str(e)
        )


# ---------------- Similar Products ----------------
@router.post("/similar")
async def similar_products(
        image: UploadFile = File(...),
        top_k: int = 5
):
    """
    Find top K similar products.
    """

    try:

        result = await image_service.find_similar_products(
            image=image,
            top_k=top_k
        )

        return {
            "success": True,
            "top_k": top_k,
            "products": result
        }

    except Exception as e:

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )


# ---------------- Reverse Image Search ----------------
@router.post("/reverse")
async def reverse_image_search(
        image: UploadFile = File(...)
):
    """
    Reverse image search endpoint.
    """

    try:

        result = await image_service.reverse_search(
            image
        )

        return {
            "success": True,
            "matches": result
        }

    except Exception as e:

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )


# ---------------- Image Recommendations ----------------
@router.post("/recommend")
async def image_recommendation(
        image: UploadFile = File(...),
        top_k: int = 10
):
    """
    Recommend products based on uploaded image.
    """

    try:

        result = await image_service.recommend_products(
            image=image,
            top_k=top_k
        )

        return {
            "success": True,
            "recommendations": result
        }

    except Exception as e:

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )


# ---------------- Health Check ----------------
@router.get("/health")
def health():

    return {
        "status": "healthy",
        "service": "image-search"
    }